### ISSUE TITLE
Any plan to release original code?

### ISSUE THREAD

**Issue by WangHexie** [HarrialX/GraphBackdoor#4]
There are serious bugs in the code.
1. The GCN model to generate adjacency matrix in trigger generation function is unable to train. The model is detached from the computation graph, so no gradient can be calculated and model is not updated at all. Related [code](https://github.com/HarrialX/GraphBackdoor/commit/36e51e876eebb5928d1643ab051cbd12963746fa#diff-eab0d3e89e9b4babc141513fb00b8d5383646b0f81ac1ff0d377ab8261268c1cR53)
2. The GCN model to generate adjacency matrix in trigger generation function has design problem. But the realization in the paper is correct. In the paper,$Z = GCN(X_0, A_0)$ $A = ZZ^T$,  but in the code $A = GCN(X_0, A_0)$. Though the shape is correct, but the output of GCN model **cannot** be used as adjacency matrix directly(Permute the row of X_0, and make the corresponding change to the A_0, you will find why).
3. None of the experiment results can be produced from the code.
4. I think the attack to the pretrained model is the important part of the paper, but why the related code is totally not included?

**Comment by zhaohan-xi** [comment]
Hi There,

(1) we don't make an end-to-end optimization to change its original adj directly; instead, we optimize a new adj then patch it into the original one. The reason we keep the original adj untouched is using it as the input into further optimizations.
(2) when generating new adjacencies, we have the implementation as shown [here](https://github.com/HarrialX/GraphBackdoor/blob/36e51e876eebb5928d1643ab051cbd12963746fa/trojan/GTA.py#L89)
(3)(4) if you read our work carefully, you may find our reported results only use the AIDS dataset in transfer learning settings. Here we release a lightweight version without the transfer setting since we're using them for other developments, hence keeping them confidentially. But if you want to defend our attacks or develop advanced backdoors to outperform our works, feel free to simply use the release codes as your base.

Btw, please make careful assertions next time before clearly and carefully going over everything :)

**Comment by WangHexie** [comment]
> (1) we don't make an end-to-end optimization to change its original adj directly; instead, we optimize a new adj then patch it into the original one. The reason we keep the original adj untouched is using it as the input into further optimizations.     

In the this version of code, the topology generator is totally not trained at all. I checked gradient and compared the models in the [function](https://github.com/HarrialX/GraphBackdoor/blob/36e51e876eebb5928d1643ab051cbd12963746fa/trojan/GTA.py#L97) before and after training. 
![image](https://user-images.githubusercontent.com/31768052/150669516-53ec32a7-744a-4bb6-acce-1a2a38b4f4e2.png)
I'm quite confident about this. This bug is caused by your change in this claim.
>GNNs: now we use DGL-based framework to implement our GNN, which has better memory occupation and running speed. For more information about DGL, see Useful resources.
  


> (2) when generating new adjacencies, we have the implementation as shown [here](https://github.com/HarrialX/GraphBackdoor/blob/36e51e876eebb5928d1643ab051cbd12963746fa/trojan/GTA.py#L89)
     
`bkdmat = torch.div(torch.add(bkdmat, bkdmat.transpose(0, 1)), 2.0)
`
The code in your link cannot realize formula in your paper 
![image](https://user-images.githubusercontent.com/31768052/150669624-c4a7b2f2-a7ec-4900-b1ae-cf0aaa757091.png)
This is the code of $A = 1/2(A' + A'^T)$
You add it directly to the initial trigger in [training](https://github.com/HarrialX/GraphBackdoor/blob/36e51e876eebb5928d1643ab051cbd12963746fa/trojan/GTA.py#L148) and [evaluation](https://github.com/HarrialX/GraphBackdoor/blob/36e51e876eebb5928d1643ab051cbd12963746fa/main/attack.py#L118). 

I've read your code for nearly one month and sorry for my tone.
